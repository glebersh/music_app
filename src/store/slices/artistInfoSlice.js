import { createSlice } from "@reduxjs/toolkit";
import { createAsyncThunk } from "@reduxjs/toolkit";

export const getArtistGeneral = createAsyncThunk(
  'artist/getArtistGeneral',
  async function (artistID, { dispatch, getState, rejectWithValue }) {
    try {
      const state = getState();
      await fetch(`https://api.spotify.com/v1/artists/${artistID}`,
        {
          headers: {
            Authorization: `Bearer ${state.authReducer}`,
          }
        })
        .then(response => response.json())
        .then(response => dispatch(setArtist(response)))
        .then(() => fetch(`https://api.spotify.com/v1/artists/${artistID}/top-tracks?market=US`,
          {
            headers: {
              Authorization: `Bearer ${state.authReducer}`,
            }
          }))
        .then(response => response.json())
        .then(response => dispatch(setArtistTopTracks(response)))
        .then(() => fetch(`https://api.spotify.com/v1/artists/${artistID}/albums?market=US&limit=10`,
          {
            headers: {
              Authorization: `Bearer ${state.authReducer}`,
            }
          }))
        .then(response => response.json())
        .then(response => dispatch(setArtistTopAlbums(response)))
        .catch(error => console.log(error.message))
    } catch (error) {
      return rejectWithValue(error.message);
    }
  });


const artistInfoSlice = createSlice({
  name: 'artistInfo',
  initialState: {
    artistInfo: {},
    artistTopTracks: {},
    artistTopAlbums: {},
    loadingStatus: '',
    errorStatus: null,
  },
  reducers: {
    setArtist(state, action) {
      state.artistInfo = action.payload;
    },
    setArtistTopTracks(state, action) {
      state.artistTopTracks = action.payload;
    },
    setArtistTopAlbums(state, action) {
      state.artistTopAlbums = action.payload;
    }
  },
  extraReducers: {
    [getArtistGeneral.pending]: (state) => {
      state.loadingStatus = 'loading';
    },
    [getArtistGeneral.fulfilled]: (state) => {
      state.loadingStatus = 'resolved';
    },
    [getArtistGeneral.rejected]: (state, action) => {
      state.errorStatus = action.payload;
      state.loadingStatus = 'rejected';
    },
  }
});

export const {
  setArtist,
  setLoading,
  setArtistTopTracks,
  setArtistTopAlbums } = artistInfoSlice.actions;

export default artistInfoSlice.reducer;