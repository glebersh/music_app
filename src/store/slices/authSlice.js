import { createSlice } from "@reduxjs/toolkit";

const authSlice = createSlice({
  name: 'authToken',
  initialState: '',
  reducers: {
    setAuthToken(_, action) {
      return action.payload;
    }
  }
});
export const { setAuthToken } = authSlice.actions;
export default authSlice.reducer;

const authSelector = (state) => state.authReducer; 