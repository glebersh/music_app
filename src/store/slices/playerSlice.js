import { createSlice } from "@reduxjs/toolkit";
import { createAsyncThunk } from "@reduxjs/toolkit";


export const getSongCollectionTracks = createAsyncThunk(
  'playerData/getSongCollectionTracks',
  async function ({ id, playlistType }, { dispatch, getState, rejectWithValue }) {
    try {
      const state = getState();
      const repsonse = await fetch(`https://api.spotify.com/v1/${playlistType}/${id}?market=US`,
        {
          headers: {
            Authorization: `Bearer ${state.authReducer}`,
          }
        })
      const data = await repsonse.json();
      dispatch(setSongsCollectionType(playlistType));
      return data;
    } catch (error) {
      return rejectWithValue(error.message);
    }
  });


const playerSlice = createSlice({
  name: 'playerData',
  initialState: {
    songsCollection: {},
    songsCollectionType: 'songs',
    isPlaying: false,
    currentSong: {},
    previousSong: {},
    nextSong: {},
    loadingStatus: '',
    errorStatus: null,
  },
  reducers: {
    setSongsCollection(state, action) {
      state.songsCollection = action.payload;
      state.loadingStatus = 'resolved';
    },
    setSongsCollectionType(state, action) {
      state.songsCollectionType = action.payload;
    },
    setIsPlaying(state) {
      state.isPlaying = !state.isPlaying;
    },
    setIsPlayingTrue(state) {
      state.isPlaying = true;
    },
    setCurrentSong(state, action) {
      state.currentSong = action.payload;
      state.loadingStatus = 'resolved';
    },
    setPreviousSong(state, action) {
      state.previousSong = action.payload;
    },
    setNextSong(state, action) {
      state.nextSong = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getSongCollectionTracks.pending, (state, action) => {
        state.loadingStatus = 'loading';
      })
      .addCase(getSongCollectionTracks.fulfilled, (state, action) => {
        state.songsCollection = action.payload;
        state.currentSong = state.songsCollectionType === 'albums'
          || state.songsCollectionType === 'playlists' ?
          state.songsCollection?.tracks?.items[0] : state.currentSong;
        state.loadingStatus = 'resolved';
      })
      .addCase(getSongCollectionTracks.rejected, (state, action) => {
        state.errorStatus = action.payload;
        state.loadingStatus = 'rejected';
      })
  }
});

export const { setSongsCollection,
  setSongsCollectionType,
  setIsPlaying,
  setIsPlayingTrue,
  setCurrentSong,
  setPreviousSong,
  setNextSong } = playerSlice.actions;

export default playerSlice.reducer;