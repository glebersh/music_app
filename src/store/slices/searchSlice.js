import { createSlice } from "@reduxjs/toolkit";
import { createAsyncThunk } from "@reduxjs/toolkit";

export const getDataFromSearch = createAsyncThunk(
  'artist/getDataFromSearch',

  async function (urlFormattedText, { dispatch, getState }) {
    const state = getState();
    await fetch(`https://api.spotify.com/v1/search?q=${urlFormattedText}&type=artist,track,album,playlist&market=US&limit=15`,
      {
        headers: {
          Authorization: `Bearer ${state.authReducer}`,
        }
      })
      .then(response => response.json())
      .then(response => dispatch(onSearch(response)))
      .catch(error => console.log(error.message));
  });


const searchSlice = createSlice({
  name: 'spotify',
  initialState: {
    searchResult: {},
    isEmpty: true,
    searchCategory: 'multi',
    isLoading: true,
  },
  reducers: {
    onSearch(state, action) {
      state.searchResult = action.payload;
      state.isEmpty = false;
      state.isLoading = false;
    },
    onLoading(state) {
      state.isLoading = true;
    },
    onCategoryChange(state, action) {
      state.searchCategory = action.payload;
    }
  }
});

export const { onSearch, onCategoryChange, onLoading } = searchSlice.actions;
export default searchSlice.reducer;