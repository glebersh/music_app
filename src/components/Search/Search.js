import React, { useState, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import './Search.css';

import { getDataFromSearch, onCategoryChange } from '../../store/slices/searchSlice';

import { Box, Flex, Input, Tag } from '@chakra-ui/react';
import { useNavigate } from 'react-router-dom';

const Search = () => {
  const dispatch = useDispatch();
  const [searchText, setSearchText] = useState();
  const [searchCategory, setSearchCategory] = useState('multi');

  useEffect(() => {
    dispatch(onCategoryChange(searchCategory));
  }, [searchCategory]);

  const navigate = useNavigate();

  const sendSearchRequest = (e, text, category) => {
    e.preventDefault();
    const urlFormattedText = text.replace(' ', '%20');
    dispatch(getDataFromSearch(urlFormattedText));
    dispatch(onCategoryChange(category));
    navigate('/');
  };

  const hoverStyle = {
    transition: '.33s',
    cursor: 'pointer',
    color: 'primary'
  };

  return (
    <Box w='100%'>
      <form onSubmit={(e) => sendSearchRequest(e, searchText, searchCategory)}>
        <Flex direction='row' flexWrap='wrap' justify='flex-start' align='center'>
          <Input type='search' onChange={(e) => setSearchText(e.target.value)}
            display='inline-block' w='50%'
            role='search' />
          <Input type='submit' value='Search'
            role='search-submit-button'
            _hover={hoverStyle}
            w='10%' minW='100px'
            ml='2em'
            display='inline-block' />
        </Flex>
        <Flex direction='row' flexWrap='wrap' justify='flex-start' align='center'
          mt='1.5em' gap='1em'>
          <Tag onClick={() => setSearchCategory('multi')}
            transition='.33s'
            _hover={hoverStyle}
            _active={{ transform: 'scale(0.97,0.97)' }}
            role='search-filter-button'>Multi</Tag>
          <Tag onClick={() => setSearchCategory('songs')}
            transition='.33s'
            _hover={hoverStyle}
            _active={{ transform: 'scale(0.97,0.97)' }}
            role='search-filter-button'>Songs</Tag>
          <Tag onClick={() => setSearchCategory('albums')}
            transition='.33s'
            _hover={hoverStyle}
            _active={{ transform: 'scale(0.97,0.97)' }}
            role='search-filter-button'>Albums</Tag>
          <Tag onClick={() => setSearchCategory('artists')}
            transition='.33s'
            _hover={hoverStyle}
            _active={{ transform: 'scale(0.97,0.97)' }}
            role='search-filter-button'>Artists</Tag>
          <Tag onClick={() => setSearchCategory('playlists')}
            _hover={hoverStyle}
            _active={{ transform: 'scale(0.97,0.97)' }}
            role='search-filter-button'>Playlists</Tag>
        </Flex>
      </form>
    </Box >
  )
};

export default Search;