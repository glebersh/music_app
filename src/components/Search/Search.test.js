import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import * as reduxHooks from 'react-redux';
import Search from './Search';
import * as searchActions from '../../store/slices/searchSlice';
import * as routerHooks from 'react-router-dom';

jest.mock('react-redux');
const mockUseDispatch = jest.spyOn(reduxHooks, 'useDispatch');
const mockedUseSelector = jest.spyOn(reduxHooks, 'useSelector');

jest.mock('react-router-dom');
const mockedUsedNavigate = jest.spyOn(routerHooks, 'useNavigate');

jest.mock('../../store/slices/searchSlice');
const mockedChangeCategory = jest.spyOn(searchActions, 'onCategoryChange');
const mockedGetDataFromSearch = jest.spyOn(searchActions, 'getDataFromSearch');

describe('Searchbar tests', () => {
  it('Should render and match snapshot', () => {
    const dispatch = jest.fn();
    mockUseDispatch.mockReturnValue(dispatch);
    const searchbar = render(<Search />);
    expect(searchbar).toMatchSnapshot();
  });

  it('Should dispatch action on tag click', () => {
    const dispatch = jest.fn();
    mockUseDispatch.mockReturnValue(dispatch);

    const searchbar = render(<Search />);

    userEvent.click(searchbar.getAllByRole('search-filter-button')[1]);
    expect(mockedChangeCategory).toHaveBeenCalledWith('songs');

    userEvent.click(searchbar.getAllByRole('search-filter-button')[2]);
    expect(mockedChangeCategory).toHaveBeenCalledWith('albums');

    expect(dispatch).toHaveBeenCalledTimes(3);
  });

  it('Should submit form', () => {
    const dispatch = jest.fn();
    mockUseDispatch.mockReturnValue(dispatch);
    const navigate = jest.fn();
    mockedUsedNavigate.mockReturnValue(navigate);

    const searchbar = render(<Search />);
    userEvent.type(searchbar.getByRole('search'), 'Red velvet');
    userEvent.click(searchbar.getByRole('search-submit-button'));

    expect(dispatch).toHaveBeenCalledTimes(3);
    expect(mockedChangeCategory).toHaveBeenCalledWith('multi');
    expect(mockedGetDataFromSearch).toHaveBeenCalledWith('Red%20velvet');
    // expect(mockedUsedNavigate).toHaveBeenCalledWith('/');
  });
});