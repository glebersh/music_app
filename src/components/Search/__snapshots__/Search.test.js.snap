// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbar tests Should render and match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-8atqhb"
      >
        <form>
          <div
            class="css-zqq8sv"
          >
            <input
              class="chakra-input css-15u1lqs"
              role="search"
              type="search"
            />
            <input
              class="chakra-input css-hn1347"
              role="search-submit-button"
              type="submit"
              value="Search"
            />
          </div>
          <div
            class="css-1hs5feh"
          >
            <span
              class="css-13tvnpb"
              role="search-filter-button"
            >
              Multi
            </span>
            <span
              class="css-13tvnpb"
              role="search-filter-button"
            >
              Songs
            </span>
            <span
              class="css-13tvnpb"
              role="search-filter-button"
            >
              Albums
            </span>
            <span
              class="css-13tvnpb"
              role="search-filter-button"
            >
              Artists
            </span>
            <span
              class="css-1qcgofx"
              role="search-filter-button"
            >
              Playlists
            </span>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-8atqhb"
    >
      <form>
        <div
          class="css-zqq8sv"
        >
          <input
            class="chakra-input css-15u1lqs"
            role="search"
            type="search"
          />
          <input
            class="chakra-input css-hn1347"
            role="search-submit-button"
            type="submit"
            value="Search"
          />
        </div>
        <div
          class="css-1hs5feh"
        >
          <span
            class="css-13tvnpb"
            role="search-filter-button"
          >
            Multi
          </span>
          <span
            class="css-13tvnpb"
            role="search-filter-button"
          >
            Songs
          </span>
          <span
            class="css-13tvnpb"
            role="search-filter-button"
          >
            Albums
          </span>
          <span
            class="css-13tvnpb"
            role="search-filter-button"
          >
            Artists
          </span>
          <span
            class="css-1qcgofx"
            role="search-filter-button"
          >
            Playlists
          </span>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
